src/libstd/sync/mutex.rs

```diff
@@ -358,6 +358,37 @@ impl<T: ?Sized> Mutex<T> {
         let data = unsafe { &mut *self.data.get() };
         poison::map_result(self.poison.borrow(), |_| data )
     }
+
+    /// Invoke a function under a mutex.
+    ///
+    /// This takes the mutex and once held passes a mutable reference to the
+    /// contained value to a `FnOnce` closure. Once the closure returns, the
+    /// mutex is released.
+    ///
+    /// # Errors
+    ///
+    /// If another user of this mutex panicked while holding the mutex, then
+    /// this call will return an error instead.
+    ///
+    /// # Example
+    /// ```
+    /// use std::sync::Mutex;
+    ///
+    /// let mutex = Mutex::new(0);
+    ///
+    /// // Atomically fetch the old value and increment it.
+    /// let old = mutex.with(|v| { let old = *v; *v += 1; old }).unwrap();
+    ///
+    /// assert_eq!(old, 0);
+    /// assert_eq!(*mutex.lock().unwrap(), 1);
+    /// ```
+    #[unstable(feature = "mutex_with", issue = "61974")]
+    pub fn with<U>(
+        &self,
+        func: impl FnOnce(&mut T) -> U,
```

Comment:
```suggestion
        func: F
```

Reply:
Is there a reason to avoid `impl` here?

Reply:
We don't use it anywhere else in the public interfaces of the standard library to my knowledge. The reason why is that when you use `impl Trait` you cannot use turbofish for `U`.